### PR TITLE
with_retries instead of retry

### PR DESCRIPTION
**Pull request by abelcastilloavant**
The code is not mine, it is @davidboren's.

### CODE REVIEW

File: R/retry.R [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+#' @param num_tries numeric. The number of attempts.
+#' @param sleep_secs numeric. How long to wait between attempts.
+#' @param expr expression. The expression to evaluate with retries.
+#' @return TRUE if the expression gets evaluated successfully on some attempt.
+#' @export
+with_retries <- function(expr, num_tries = 1, sleep_secs = 0.001) {
+  num_tries <- max(num_tries, 1L)
+  current_try <- 0
+  while (current_try < num_tries) {
+    out <- evaluate::try_capture_stack(expr, env = parent.frame())
```
Comment (peterhurford) [inline R/retry.R:12]: Is `evaluate` a base package or does it need to be in `DESCRIPTION`? What's the advantage of `try_capture_stack` over `try`?

File: R/retry.R [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
+    if (!is(out, "error")) break
```
Comment (peterhurford) [inline R/retry.R]: Can you prefix with `methods::`?

File: R/retry.R [hunk 3]
```diff
@@ -1,10 +1,27 @@
...
-  tries <- 0
-  repeat {
-    tries <- tries + 1
-    if (tries >= times) { break }
-    outcome <- try(eval(expr), silent = TRUE)
-    if (methods::is(outcome, "try-error")) { next } else { break }
+#' Try evaluating an expression multiple times before erroring.
+#'
+#' @param num_tries numeric. The number of attempts.
+#' @param sleep_secs numeric. How long to wait between attempts.
```
Comment (peterhurford) [inline R/retry.R]: Can we call this `sleep`?

File: R/retry.R [hunk 4]
```diff
@@ -1,10 +1,27 @@
...
+#' @param expr expression. The expression to evaluate with retries.
+#' @return TRUE if the expression gets evaluated successfully on some attempt.
+#' @export
```
Comment (peterhurford) [inline R/retry.R:7]: Can you re-run document?

File: R/retry.R [hunk 5]
```diff
@@ -1,10 +1,27 @@
...
+with_retries <- function(expr, num_tries = 1, sleep_secs = 0.001) {
+  num_tries <- max(num_tries, 1L)
+  current_try <- 0
+  while (current_try < num_tries) {
+    out <- evaluate::try_capture_stack(expr, env = parent.frame())
+    if (!is(out, "error")) break
+    current_try <- current_try + 1
+    message(paste0('Attempt ', current_try, ' failed.'))
```
Comment (peterhurford) [inline R/retry.R]: Can you use double quotes?

File: R/retry.R [hunk 6]
```diff
@@ -1,10 +1,27 @@
...
+    if (current_try < num_tries) {
+      cat(' Sleeping for ', sleep_secs, 'seconds\n')
```
Comment (peterhurford) [inline R/retry.R]: I don't think you want the beginning space?

File: R/retry.R [hunk 7]
```diff
@@ -1,10 +1,27 @@
...
+      Sys.sleep(sleep_secs)
+    }
+  }
+  if (is(out, "error")) {
```
Comment (peterhurford) [inline R/retry.R]: `methods::is`.

File: tests/testthat/test-retry.R [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    expect_equal(x, "didn't error")
+  })
+
+  test_that("it only retries the prescribed number of times", {
+    attempt_number <<- 1
+    expect_error(
+      with_retries({ x <- pass_on_third_attempt() }, num_tries = 2, sleep_secs = 0),
+      "not enough retries"
+    )  
+  })
```
Comment (peterhurford) [inline tests/testthat/test-retry.R:27]: Nice tests 👍

File: R/retry.R (hunk 1 above)
Comment (abelcastilloavant) [inline R/retry.R:12]: From conversation with @davidboren:

> Probably will have to throw evaluate into the suggests
> the evaluate package is the only reason we can extract the stacktrace.  Super handy

**Comment by peterhurford** [comment]
LGTM :+1: :shipit:

**Pull request merged**